src/drivers/intel/dmic.c

```diff
@@ -1620,18 +1620,16 @@ static int dmic_remove(struct dai *dai)
 
 	dai_info(dai, "dmic_remove()");
 
-	/* remove scheduling */
-	schedule_task_free(&dmic->dmicwork);
-
-	rfree(dai_get_drvdata(dai));
-	dai_set_drvdata(dai, NULL);
-
 	interrupt_disable(dmic->irq, dai);
 	interrupt_unregister(dmic->irq, dai);
 
+	/* remove scheduling */
+	schedule_task_free(&dmic->dmicwork);
+
 	dai_info(dai, "dmic_remove(), dmic_active_fifos_mask = 0x%x, dmic_pause_mask = 0x%x",
 		 active_fifos_mask, pause_mask);
 	rfree(dmic);
+	dai_set_drvdata(dai, NULL);
```

Comment:
to be pedantic these two lines should be swapped: first make sure nobody can reach driver data, then free it, but we're safe here - all asynchronous flows are stopped already

Reply:
ack fixed!